[Feature-8612][RESOURCE]  extend s3 to the storage of ds

nobolity opened the pull request:
1.fix some spell question
2.extend the type of storage
3.add the s3utils
to manager resource
4.automatic inject the storage in addition to your
config

<!--Thanks very much for contributing to Apache DolphinScheduler. Please review https://dolphinscheduler.apache.org/en-us/community/development/pull-request.html before opening a pull request.-->


## Purpose of the pull request
close #8612
<!--(For example: This pull request adds checkstyle plugin).-->

## Brief change log

<!--*(for example:)*
  - *Add maven-checkstyle-plugin to root pom.xml*
-->
## Verify this pull request

<!--*(Please pick either of the following options)*-->

This pull request is code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:

<!--*(example:)*
  - *Added dolphinscheduler-dao tests for end-to-end.*
  - *Added CronUtilsTest to verify the change.*
  - *Manually verified the change by testing locally.* -->

nobolity committed:
feat(resource  manager): extend s3 to the storage of ds

1.fix some spell question
2.extend the type of storage
3.add the s3utils
to manager resource
4.automatic inject the storage in addition to your
config

CalvinKirs commented on dolphinscheduler-api/pom.xml:
```diff
@@ -245,6 +245,11 @@
             </exclusions>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>mysql</groupId>
+            <artifactId>mysql-connector-java</artifactId>
+            <scope>compile</scope>
+        </dependency>
```
we don't use mysql in compile. and why do you add this?

nobolity committed:
fix(resource  manager): update the dependency

zhongjiajie commented on dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/ResourcesController.java:
```diff
@@ -496,6 +464,7 @@ public Result createUdfFunc(@ApiIgnore @RequestAttribute(value = Constants.SESSI
                                 @RequestParam(value = "database", required = false) String database,
                                 @RequestParam(value = "description", required = false) String description,
                                 @PathVariable(value = "resourceId") int resourceId) {
+        //todo verify the sourceName
```
Why we need `todo` here? Does this WIP?

nobolity replied:
we  verify fewer parameters from the controller 
eg the name can't include '/'
so we need the rule which hasn`t been designed

zhongjiajie reviewed (commented):
Does this PR still WIP? I find some `todo` and meanless comment here

zhongjiajie commented on dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/configuration/StoreConfiguration.java:
```diff
@@ -0,0 +1,54 @@
+import static org.apache.dolphinscheduler.common.Constants.*;
+
+/**
+ * @author Terry.Dai
```
I think we should remove `author` attribute

nobolity replied:
had removed it

zhongjiajie commented on dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/ResourcesController.java:
```diff
@@ -49,41 +30,25 @@
 
+import java.util.Map;
+
+import static org.apache.dolphinscheduler.api.enums.Status.*;
```
Should be avoid import *

nobolity replied:
had  merged it

nobolity committed:
fix(resource  manager): extend s3 to the storage of ds

fix the constant of hadooputils

nobolity committed:
fix(resource  manager): extend s3 to the storage of ds

1.fix some spell question
2.delete the import *

nobolity committed:
fix(resource  manager):

merge  the unitTest:
1.TenantServiceImpl
2.ResourceServiceImpl
3.UserServiceImpl

nobolity committed:
fix(resource  manager): extend s3 to the storage of ds

merge the resourceServiceTest

nobolity committed:
fix(resource  manager): test  cancel the test method

createTenant verifyTenant

nobolity committed:
fix(resource  manager): merge the code  follow the check-result of sonar

nobolity committed:
fix(resource  manager): extend s3 to the storage of ds

fit the spell question

nobolity committed:
fix(resource  manager): extend s3 to the storage of ds

revert the common.properties

nobolity committed:
Merge remote-tracking branch 'upstream/dev' into dev

nobolity committed:
fix(resource  manager): extend s3 to the storage of ds

update the storageConfig with None

nobolity committed:
fix(resource  manager): extend s3 to the storage of ds

fix the judge of resourceType

nobolity committed:
fix(resource  manager): extend s3 to the storage of ds

undo the compile-mysql

nobolity committed:
fix(resource  manager): extend s3 to the storage of ds

delete hadoop aws

nobolity committed:
fix(resource  manager): extend s3 to the storage of ds

update the know-dependencies to delete aws 1.7.4
update the e2e
file-manager common.properties

nobolity committed:
fix(resource  manager): extend s3 to the storage of ds

update the aws-region

nobolity committed:
fix(resource  manager): extend s3 to the storage of ds

fix the storageconfig init

nobolity committed:
fix(resource  manager): update e2e docker-compose

update e2e docker-compose

nobolity committed:
fix(resource  manager): extend s3 to the storage of ds

revent the e2e common.proprites

print the resource type in propertyUtil

nobolity committed:
fix(resource  manager): extend s3 to the storage of ds
1.println the properties

nobolity committed:
fix(resource  manager): println the s3 info

nobolity committed:
fix(resource  manager): extend s3 to the storage of ds

delete the info  and upgrade the s3 info to e2e

caishunfeng commented on dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/BaseServiceImpl.java:
```diff
@@ -37,6 +35,7 @@
 
 /**
  * base service impl
+ * @author Terry.Dai
```
Please remove it.

nobolity replied:
had removed it

caishunfeng commented on dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ResourcesServiceImpl.java:
```diff
@@ -756,20 +749,19 @@ private boolean upload(User loginUser, String fullName, MultipartFile file, Reso
-                    if (HadoopUtils.getInstance().exists(hdfsFilename)) {
-                        logger.error("resource type:{} name:{} has exist in hdfs {}, can't create again.", type, RegexUtils.escapeNRT(fullName), hdfsFilename);
-                        putMsg(result, Status.RESOURCE_FILE_EXIST,hdfsFilename);
+                    //xxx
```
remove it.

nobolity replied:
had removed it

caishunfeng commented on dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/UsersServiceImpl.java:
```diff
@@ -28,54 +32,23 @@
-import org.apache.dolphinscheduler.dao.mapper.TenantMapper;
-import org.apache.dolphinscheduler.dao.mapper.UDFUserMapper;
-import org.apache.dolphinscheduler.dao.mapper.UserMapper;
+import org.apache.dolphinscheduler.dao.entity.*;
```
please avoid to import *

caishunfeng commented on dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/UsersServiceImpl.java:
```diff
@@ -462,34 +433,32 @@ public Result queryUserList(User loginUser, String searchVal, Integer pageNo, In
+            } catch (Exception e) {
+                logger.error("create tenant {} failed ,the reason is {}", newTenant, e.getMessage());
+            }
+            }
```
code format?

caishunfeng commented:
Hi @nobolity Can you show the brief change log about the key code in pr description?
BTW, this feature shouled be docemented.

nobolity committed:
fix(resource  manager): extend s3 to the storage of ds

add the bucket init

nobolity committed:
fix(resource  manager): extend s3 to the storage of ds

1.fix some spell question
2.delete the import *

SbloodyS commented on dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/utils/S3Utils.java:
```diff
@@ -0,0 +1,280 @@
+            s3Client = AmazonS3ClientBuilder
+                    .standard()
+                    .withCredentials(new AWSStaticCredentialsProvider(new BasicAWSCredentials(ACCESS_KEY_ID, SECRET_KEY_ID)))
+                    .withRegion(Regions.fromName(REGION)).build();
```
This may lead to ```AmazonS3Exception``` while using ```Minio S3``` and ```endpoint``` using domain name to connect.

Because AWS SDK has some conversion rules when using domain name links. Related articles： https://aws.amazon.com/cn/blogs/aws/amazon-s3-path-deprecation-plan-the-rest-of-the-story/

You may use ```.withPathStyleAccessEnabled(true)``` to avoid this problems. 
Related articles：https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/s3/AmazonS3Builder.html#withPathStyleAccessEnabled-java.lang.Boolean-

nobolity committed:
fix(resource  manager): extend s3 to the storage of ds

upgrade the s3 endpoint

nobolity committed:
Merge remote-tracking branch 'upstream/dev' into dev

nobolity committed:
fix(resource  manager): withPathStyleAccessEnabled(true)

nobolity committed:
fix(resource  manager): extend s3 to the storage of ds

1.fix some spell question
2.delete the import *

nobolity committed:
fix(resource  manager): upgrade the  s3client builder

nobolity committed:
fix(resource  manager): correct  the s3 point to s3client

nobolity committed:
fix(resource  manager): update the constant BUCKET_NAME

nobolity committed:
fix(resource  manager): e2e  s3 endpoint -> s3:9000

nobolity committed:
Merge remote-tracking branch 'upstream/dev' into dev

nobolity committed:
fix(resource  manager): extend s3 to the storage of ds

1.fix some spell question
2.delete the import *

zhongjiajie commented:
I restart the failed CI

nobolity committed:
Merge remote-tracking branch 'upstream/dev' into dev

zhongjiajie commented:
I approve to run now

zhongjiajie commented:
@nobolity Hi, I think these errors in `E2E / UdfManage` related to this PR

```log
Error:  Errors: 
[1419](https://github.com/apache/dolphinscheduler/runs/5432210214?check_suite_focus=true#step:6:1419)
Error:    UdfManageE2ETest.testCreateDirectory:120 » ConditionTimeout Assertion conditio...
[1420](https://github.com/apache/dolphinscheduler/runs/5432210214?check_suite_focus=true#step:6:1420)
Error:    UdfManageE2ETest.testDeleteDirectory:148 » Runtime No delete button in udf man...
[1421](https://github.com/apache/dolphinscheduler/runs/5432210214?check_suite_focus=true#step:6:1421)
Error:    UdfManageE2ETest.testDeleteUdf:219 » Runtime No delete button in udf manage li...
[1422](https://github.com/apache/dolphinscheduler/runs/5432210214?check_suite_focus=true#step:6:1422)
Error:    UdfManageE2ETest.testRenameDirectory:131 » Runtime No rename button in udf man...
[1423](https://github.com/apache/dolphinscheduler/runs/5432210214?check_suite_focus=true#step:6:1423)
Error:    UdfManageE2ETest.testRenameUdf:204 » Runtime No rename button in udf manage li...
[1424](https://github.com/apache/dolphinscheduler/runs/5432210214?check_suite_focus=true#step:6:1424)
Error:    UdfManageE2ETest.testUploadUdf:169 » NoSuchElement no such element: Unable to ...
```

nobolity committed:
Merge remote-tracking branch 'upstream/dev' into dev

nobolity committed:
Merge remote-tracking branch 'upstream/dev' into dev

nobolity committed:
style(resource  manager): add info to createBucket

nobolity committed:
style(resource  manager): debug the log

nobolity committed:
ci(resource  manager): test

test s3

nobolity committed:
ci(ci): add INSERT INTO dolphinscheduler.t_ds_tenant (id, tenant_code, description, queue_id, create_time, update_time) VALUES(1, 'root', NULL, 1, NULL, NULL); to h2.sql

nobolity committed:
fix(resource  manager): update the h2 sql

nobolity committed:
fix(resource  manager): solve to delete the tenant

nobolity committed:
style(resource  manager): merge the style end delete the unuse s3 config

nobolity committed:
Merge remote-tracking branch 'upstream/dev' into dev

nobolity committed:
Merge remote-tracking branch 'upstream/dev' into dev

# Conflicts:
#	dolphinscheduler-service/src/main/java/org/apache/dolphinscheduler/service/quartz/impl/QuartzExecutorImpl.java

nobolity committed:
Merge remote-tracking branch 'upstream/dev' into dev

# Conflicts:
#	dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/AccessTokenServiceImpl.java
#	dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/UsersServiceImpl.java
#	dolphinscheduler-common/pom.xml
#	dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/utils/HadoopUtils.java
#	dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-api/src/main/java/org/apache/dolphinscheduler/plugin/datasource/api/utils/CommonUtils.java
#	dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/task/BaseTaskProcessor.java
#	dolphinscheduler-standalone-server/src/main/assembly/dolphinscheduler-standalone-server.xml
#	dolphinscheduler-worker/src/main/java/org/apache/dolphinscheduler/server/worker/runner/TaskExecuteThread.java
#	dolphinscheduler-worker/src/main/java/org/apache/dolphinscheduler/server/worker/runner/WorkerManagerThread.java
#	tools/dependencies/known-dependencies.txt

nobolity committed:
Merge remote-tracking branch 'upstream/dev' into dev

nobolity committed:
fix(resource  manager): extend s3 to the storage of ds

UPDATE the rename resources when s3

nobolity committed:
fix(resource  manager): extend s3 to the storage of ds

1.fix the code style of QuartzImpl

nobolity committed:
fix(resource  manager): extend s3 to the storage of ds

1.impoort restore_type to CommonUtils

nobolity committed:
fix(resource  manager): update the work thread

nobolity committed:
fix(resource  manager): update  the baseTaskProcessor

nobolity committed:
fix(resource  manager): upgrade dolphinscheduler-standalone-server.xml

nobolity committed:
fix(resource  manager): add  user Info to dolphinscheduler_h2.sql

nobolity committed:
fix(resource  manager): merge  the resourceType to NONE

nobolity committed:
Merge remote-tracking branch 'upstream/dev' into dev

nobolity committed:
style(upgrade the log level to info):

nobolity committed:
fix(resource  manager): sysnc the h2.sql

nobolity committed:
fix(resource  manager): update the merge the user tenant

nobolity committed:
Merge remote-tracking branch 'upstream/dev' into dev

# Conflicts:
#	dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ResourcesServiceImpl.java

nobolity committed:
fix(resource  manager): merge the resourcesServiceImpl

nobolity committed:
Merge remote-tracking branch 'upstream/dev' into dev

nobolity committed:
fix(resource  manager):

when the storage is s3 ,that the directory can't be renamed

nobolity committed:
fix(resource  manager): in s3 ,the directory cannot be renamed

nobolity committed:
fix(resource  manager): delete the deleteRenameDirectory in E2E

zhongjiajie commented on dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/enums/Status.java:
```diff
@@ -297,7 +297,8 @@
     HDFS_NOT_STARTUP(60001, "hdfs not startup", "hdfs未启用"),
-
+    STORAGE_NOT_STARTUP(60002, "storage not startup", "存储未启用"),
+    STORAGE_CANNOT_RENAME(60003, "directory cannot be renamed", "S3无法重命名文件夹"),
```
Is it only for `S3` or for all type of storage? if for all storage we should not use keyword `S3`

nobolity replied:
had changed STORAGE_CANNOT_RENAME to S3_CANNOT_RENAME

zhongjiajie commented on dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/enums/Status.java:
```diff
@@ -390,6 +391,8 @@
     VERIFY_K8S_NAMESPACE_ERROR(1300007, "verify k8s and namespace error", "验证k8s命名空间信息错误"),
     DELETE_K8S_NAMESPACE_BY_ID_ERROR(1300008, "delete k8s namespace by id error", "删除命名空间错误"),
+    VERIFY_PARAMETER_NAME_FAILED(503,"The file name verify  failed","文件命名校验失败"),
+    STORE_OPERATE_CREATE_ERROR(504,"create the resource failed","存储操作失败")
```
Should we avoid using an error code the same as exists/default HTTP code?

nobolity replied:
i changed thems to 130009 and 1300010

zhongjiajie commented on dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/DqRuleServiceImpl.java:
```diff
@@ -17,63 +17,42 @@
-import com.fasterxml.jackson.annotation.JsonInclude;
-import com.fasterxml.jackson.core.JsonProcessingException;
-import com.fasterxml.jackson.databind.ObjectMapper;
+import java.util.*;
```
We should avoid using `improt *`I wonder why it could pass style check?

nobolity replied:
i had used static to replaced it

SbloodyS commented on dolphinscheduler-e2e/dolphinscheduler-e2e-case/src/test/java/org/apache/dolphinscheduler/e2e/cases/FileManageE2ETest.java:
```diff
@@ -170,22 +170,25 @@ void testCancelCreateDirectory() {
+* */
+//    @Test
+//    @Order(22)
+//    void testRenameDirectory() {
```
This should not be comment out. It's useful for not aws s3 storage type. Same as ```UdfManageE2ETest```.

nobolity replied:
had  recovered it

SbloodyS replied:
It seems still be commented out.

nobolity replied:
o,i will repaired it in the text version

kezhenxu94 replied:
Can’t we repair in this PR? If this test cannot pass in this PR I’d doubt that this PR breaks something? Is it the case?

nobolity replied:
it can be fixed,but it need a few of time ,in fact ,i need to copy all files of the old directory to the new directory ,if the number of files is large, then it is easy to wrong,
it 's better to generate the diffentenct e2eTests  depend on the storage types

caishunfeng commented on dolphinscheduler-api/pom.xml:
```diff
@@ -145,6 +151,12 @@
         <dependency>
             <groupId>io.swagger</groupId>
             <artifactId>swagger-models</artifactId>
+            <exclusions>
```
Why to exclude swagger module?

nobolity replied:
had  removed it

kezhenxu94 replied:
This is not addressed.

caishunfeng commented on dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/BaseService.java:
```diff
@@ -74,21 +73,22 @@
-     * @throws IOException if hdfs operation exception
-     */
-    void createTenantDirIfNotExists(String tenantCode) throws IOException;
+//    /**
```
remove unuse code

nobolity replied:
had removed it

caishunfeng commented on dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/UsersServiceImpl.java:
```diff
@@ -432,65 +426,63 @@ public Result queryUserList(User loginUser, String searchVal, Integer pageNo, In
+//                        if (CollectionUtils.isNotEmpty(fileResourcesList)) {
+//                            ResourceTreeVisitor resourceTreeVisitor = new ResourceTreeVisitor(fileResourcesList);
+//                            ResourceComponent resourceComponent = resourceTreeVisitor.visit();
+//                            copyResourceFiles(oldTenant.getTenantCode(), newTenantCode, resourceComponent, oldResourcePath, newResourcePath);
```
Is there any problem with commenting out this code?

nobolity replied:
in fact , resources (include file ,authority,mem,cpu,project ) are bundle by tenant, but no user
so  it's not necessary to copy the resources to another place
the user can do nothing if he has no tenantCode

kezhenxu94 replied:
If so please remove them. It’s not a good practice to have a large amount of commented codes in the repo

caishunfeng commented on dolphinscheduler-api/src/test/java/org/apache/dolphinscheduler/api/controller/TenantControllerTest.java:
```diff
@@ -37,10 +29,14 @@
     private static final Logger logger = LoggerFactory.getLogger(TenantControllerTest.class);
 
-    @Test
+    //    @Test
```
why skip this test?

nobolity replied:
i  had recovered it

caishunfeng commented on dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/config/StoreConfiguration.java:
```diff
@@ -0,0 +1,53 @@
+
+
+/**
+ * @version since 2.0.4
```
please remove it and add some description for this Class

nobolity replied:
i had  sloved it

caishunfeng commented on dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/utils/HadoopUtils.java:
```diff
@@ -162,14 +149,14 @@ private void init() {
-                configuration.set(Constants.FS_S3A_SECRET_KEY, PropertyUtils.getString(Constants.FS_S3A_SECRET_KEY));
-                fs = FileSystem.get(configuration);
-            }
+//            } else if (resUploadType == ResUploadType.S3) {
```
remove the unuse code.

nobolity replied:
i had removed it

caishunfeng commented on dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/utils/S3Utils.java:
```diff
@@ -0,0 +1,299 @@
+                        .withRegion(Regions.fromName(REGION))
+                        .build();
+            }
+            logger.info("s3client is null :{}", null == s3Client);
```
```suggestion
       if (null == s3Client) {
           logger.error("s3client is null");
       }
```

nobolity replied:
had removed it

caishunfeng commented on dolphinscheduler-worker/src/main/java/org/apache/dolphinscheduler/server/worker/runner/TaskExecuteThread.java:
```diff
@@ -300,13 +310,15 @@ private void downloadResource(String execLocalPath, Map<String, String> projectR
                 try {
                     // query the tenant code of the resource according to the name of the resource
-                    String resHdfsPath = HadoopUtils.getHdfsResourceFileName(tenantCode, fullName);
+                    //XXX
```
what's the `xxx` mean?

nobolity replied:
i  had removed it

nobolity committed:
fix(resource  manager): check the style and  recoverd the test

nobolity committed:
fix(resource  manager): delete the log.print(LoginUser)

lenboo reviewed (approved):
+1

kezhenxu94 commented on dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/BaseServiceImpl.java:
```diff
@@ -127,23 +125,23 @@ public boolean check(Map<String, Object> result, boolean bool, Status userNoOper
-        HadoopUtils.getInstance().mkdir(resourcePath);
-        HadoopUtils.getInstance().mkdir(udfsPath);
-    }
+//    @Override
```
Why removed these codes? If they are not needed I’d just remove them

kezhenxu94 commented:
@nobolity please create an issue to track the unsolved problems in this PR and fix them as soon as possible
